src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -99,33 +184,31 @@ internal static string IPv6AddressToString(ushort[] address, uint scopeId)
                 buffer.Append('%').Append(scopeId);
             }
 
-            return StringBuilderCache.GetStringAndRelease(buffer);
+            return buffer;
         }
 
         private static unsafe void FormatIPv4AddressNumber(int number, char* addressString, ref int offset)
         {
-            int i = offset;
+            offset += number > 99 ? 3 : number > 9 ? 2 : 1;
+
+            int i = 0;
             do
             {
                 int rem;
                 number = Math.DivRem(number, 10, out rem);
-                addressString[--i] = (char)('0' + rem);
+                addressString[offset - ++i] = (char)('0' + rem);
             } while (number != 0);
-            offset = i;
+
         }
 
-        public static unsafe bool Ipv4StringToAddress(string ipString, out long address)
+        public static unsafe bool Ipv4StringToAddress(char* ipStringPtr, int ipStringPtrLength, out long address)
```

Comment:
As earlier, it'd be good to using `ReadOnlySpan<char>` instead of the original `string`, rather than substituting `char*`.  Especially in this IPAddress code, it'd be great to try to avoid adding any unnecessary unsafe code.